mcstatus/protocol/connection.py

```diff
@@ -20,199 +22,501 @@
 
-    def flush(self) -> bytearray:
-        result = self.sent
-        self.sent = bytearray()
-        return result
+    def write_varint(self, value: int) -> None:
+        """
+        Write varint with value value to self.
+
+        Max: 2 ** 31 - 1, Min: -(2 ** 31).
+        Raises ValueError if varint is out of range.
+        """
+        remaining = unsigned_int32(value).value
+        for _ in range(5):
+            if not remaining & -0x80:  # remaining & ~0x7F == 0:
+                self.write(struct.pack("!B", remaining))
+                if value > 2**31 - 1 or value < -(2**31):
+                    break
+                return
+            self.write(struct.pack("!B", remaining & 0x7F | 0x80))
+            remaining >>= 7
+        raise ValueError(f'The value "{value}" is too big to send in a varint')
+
+    def write_varlong(self, value: int) -> None:
+        """
+        Write varlong with value value to self.
+
+        Max: 2 ** 63 - 1, Min: -(2 ** 63).
+        Raises ValueError if varlong out of range.
+        """
+        remaining = unsigned_int64(value).value
+        for _ in range(10):
+            if not remaining & -0x80:  # remaining & ~0x7F == 0:
+                self.write(struct.pack("!B", remaining))
+                if value > 2**63 - 1 or value < -(2**31):
+                    break
+                return
+            self.write(struct.pack("!B", remaining & 0x7F | 0x80))
+            remaining >>= 7
+        raise ValueError(f'The value "{value}" is too big to send in a varlong')
+
+    def write_utf(self, value: str) -> None:
+        """Write varint of length of value up to 32767 bytes, then write value encoded with utf8."""
+        self.write_varint(len(value))
+        self.write(bytearray(value, "utf8"))
+
+    def write_ascii(self, value: str) -> None:
+        """Write value encoded with ISO-8859-1, then write an additional 0x00 at the end."""
+        self.write(bytearray(value, "ISO-8859-1"))
+        self.write(bytearray.fromhex("00"))
+
+    def write_short(self, value: int) -> None:
+        """Write 2 bytes for value -32768 - 32767"""
+        self.write(self._pack("h", value))
+
+    def write_ushort(self, value: int) -> None:
+        """Write 2 bytes for value 0 - 65535 (2 ** 16 - 1)."""
+        self.write(self._pack("H", value))
+
+    def write_int(self, value: int) -> None:
+        """Write 4 bytes for value -2147483648 - 2147483647."""
+        self.write(self._pack("i", value))
+
+    def write_uint(self, value: int) -> None:
+        """Write 4 bytes for value 0 - 4294967295 (2 ** 32 - 1)."""
+        self.write(self._pack("I", value))
+
+    def write_long(self, value: int) -> None:
+        """Write 8 bytes for value -9223372036854775808 - 9223372036854775807."""
+        self.write(self._pack("q", value))
+
+    def write_ulong(self, value: int) -> None:
+        """Write 8 bytes for value 0 - 18446744073709551613 (2 ** 64 - 1)."""
+        self.write(self._pack("Q", value))
```

Comment:
There is a couple of methods with duplicated logic. Maybe you should do it in one method with dict or possibly enum?
```py
from enum import Enum

class ValueType(Enum):
    FULL_NAME = "Q"
    ...

def write_value_with_type(self, value: int, type: ValueType) -> None:
    self.write(self._pack(type.value, value))

write_value_with_type(123, ValueType.FULL_NAME)
```
https://docs.python.org/3/library/enum.html

Reply:
Cool idea, but the original connection implementation was already using these methods, and this PR's goal is mainly to introduce async support, not necessarily to fix bad design patters in that code.

This change would be compatibility breaking and will require changes in the code-base and depracation periods on these old methods (although I'm not sure how many users of the library are actually using the connection class directly, I don't expect that number to be very big). I'll leave this up to the author to decide whether they're willing to implement this, if not, it can be done later in another PR.

Reply:
`mcstatus` really give so low level API? I thought it's just a backend.

I was pretty sure all public API is `server.py` file and answer objects, because only it somehow documented in README.

Reply:
Yeah you're right, deprecation wouldn't actually be necessary here. Still though, leaving this up to the author on whether they want to implement it here or not.

Reply:
That is a good idea, to avoid duplicated code, but I agree that it would be compatibility breaking. I think that would be better suited for a new PR instead of this one. This one changes things enough as is lol.

Reply:
This is now tracked in #355, marking this discussion as resolved.